air/src/stack/mod.rs

```diff
@@ -91,6 +117,142 @@ pub fn enforce_unique_constraints<E: FieldElement>(
     constraint_offset
 }
 
+/// Enforces generic constraints of all the stack ops.
+pub fn enforce_generic_constraints<E: FieldElement>(
+    frame: &EvaluationFrame<E>,
+    result: &mut [E],
+    op_flag: &op_flags::OpFlags<E>,
+) -> usize {
+    // enforces constraint on the 1st element in the stack in the next trace.
+    let flag_sum_0 = op_flag.no_change_0() + op_flag.left_shift_1();
+    let expected_next_item_0 =
+        op_flag.no_change_0() * frame.stack_item(0) + op_flag.left_shift_1() * frame.stack_item(1);
+    result[0] = are_equal(frame.stack_item_next(0) * flag_sum_0, expected_next_item_0);
+
+    // enforces constraint on the 2nd element in the stack in the next trace.
+    let flag_sum_1 = op_flag.no_change_1() + op_flag.left_shift_2() + op_flag.right_shift_0();
+    let expected_next_item_1 = op_flag.no_change_1() * frame.stack_item(1)
+        + op_flag.left_shift_2() * frame.stack_item(2)
+        + op_flag.right_shift_0() * frame.stack_item(0);
+    result[1] = are_equal(frame.stack_item_next(1) * flag_sum_1, expected_next_item_1);
```

Comment:
This works - but it results in quite a bit of code - we could probably try to simplify this using a loop. For this, we'd need to change the signature of `left_shift()`, `right_shift()`, and `no_change()` methods to accept an index (similar to how we do it with stack items). Then, the body of the loop would look something like:

```Rust
let flag_sum = op_flag.no_change(i) + op_flag.left_shift(i + 1) + op_flag.right_shift(i - 1);
let expected_next_item = op_flag.no_change(i) * frame.stack_item(i)
        + op_flag.left_shift(i + 1) * frame.stack_item(i + 1)
        + op_flag.right_shift(i - 1) * frame.stack_item(i - 1);
result[i] = are_equal(frame.stack_item_next(i) * flag_sum, expected_next_item);
```

We'd still have to do the first and the last iterations manually, but at least all other iterations would be handled by a loop.

Reply:
Yep, in this case having all arrays go up to 16 is a better option.